File: example-crs-webservice/crs-userspace/libs/libMSA/python/libmsa/kafka/consumer.py
from typing import Type, TypeVar

from confluent_kafka import Consumer as KafkaConsumer
from confluent_kafka import KafkaException
from google.protobuf.message import Message

T = TypeVar("T", bound=Message)


class Consumer:
    def __init__(
        self,
        bootstrap_server_addr: str,
        topic: str,
        group_id: str,
        protobuf_class: Type[T],
    ) -> None:
        self.bootstrap_server_addr = bootstrap_server_addr
        self.topic = topic
        self.group_id = group_id
        self.protobuf_class = protobuf_class

        self.consumer = KafkaConsumer(
            {
                "bootstrap.servers": self.bootstrap_server_addr,
                "group.id": self.group_id,
                "auto.offset.reset": "earliest",
                "enable.partition.eof": "false",
                "enable.auto.commit": "false",
                "session.timeout.ms": 6000,
                "heartbeat.interval.ms": 2000,
            }
        )
        self.consumer.subscribe([self.topic])

    def recv_message(self) -> Message | None:
        try:
            msg = self.consumer.poll(
                0.1
            )  # Poll for messages with a timeout of 1 second
            if msg is None:  # No messages available
                return None
            if msg.error():
                raise KafkaException(msg.error())

            self.consumer.commit()

            # Deserialize the Protobuf message
            deserialized_message = self.protobuf_class()
            msg_value = msg.value()
            if isinstance(msg_value, bytes):
                deserialized_message.ParseFromString(msg_value)
                return deserialized_message
            else:
                raise TypeError(
                    f"Message value cannot be a {type(msg_value)}. Expected bytes."
                )

        except Exception as e:
            print(f"Error while receiving message: {e}")
            return None

    def close(self):
        self.consumer.close()

File: example-crs-webservice/crs-userspace/libs/libMSA/rust/libmsa/src/kafka/consumer.rs
use prost::Message;
use rdkafka::config::ClientConfig;
use rdkafka::consumer::{CommitMode, Consumer as KafkaConsumer, StreamConsumer};
use rdkafka::message::Message as KafkaMessage;

pub struct Consumer<M>
where
    M: Message + std::default::Default,
{
    bootstrap_server_addr: String,
    topic: String,
    group_id: String,
    consumer: StreamConsumer,
    _marker: std::marker::PhantomData<M>,
}

impl<M> Consumer<M>
where
    M: Message + std::default::Default,
{
    pub fn new(bootstrap_server_addr: String, topic: String, group_id: String) -> Self {
        let consumer: StreamConsumer = ClientConfig::new()
            .set("group.id", group_id.clone())
            .set("bootstrap.servers", bootstrap_server_addr.clone())
            .set("enable.partition.eof", "false")
            .set("auto.offset.reset", "earliest")
            .set("enable.auto.commit", "false")
            .set("session.timeout.ms", "6000")
            .set("heartbeat.interval.ms", "2000")
            .create()
            .expect("Consumer creation failed");

        consumer
            .subscribe(&[&topic])
            .expect("Failed to subscribe to topic");

        Self {
            bootstrap_server_addr,
            topic,
            group_id,
            consumer,
            _marker: std::marker::PhantomData,
        }
    }

    pub async fn recv_message(&self) -> M {
        loop {
            match self.consumer.recv().await {
                Ok(msg) => {
                    if let Some(payload) = msg.payload() {
                        match M::decode(payload) {
                            Ok(proto_message) => {
                                self.consumer
                                    .commit_message(&msg, CommitMode::Sync)
                                    .unwrap();
                                return proto_message;
                            }
                            Err(e) => {
                                println!("Failed to decode protobuf message: {:?}", e);
                            }
                        }
                    }
                }
                Err(e) => {
                    println!("Error receiving message: {:?}", e);
                }
            }
        }
    }
}
